### PR TITLE
feat(sagemaker): add Endpoint L2 construct

### LINKED ISSUE #2809: sagemaker: Create L2 construct for Amazon SageMaker
There is currently only L1 support for Amazon SageMaker. Would be nice to have an L2 construct to support SageMaker notebook instances, models and endpoints.

This is being actively worked on in three phases 🚀:

- [x]  feat(sagemaker): add Model L2 construct
- [ ] feat(sagemaker): add EndpointConfig L2 construct
- [ ] feat(sagemaker): add Endpoint L2 construct

### PR DESCRIPTION
**Pull request by petermeansrock**
This is the third and final PR to complete the implementation of RFC 431:
https://github.com/aws/aws-cdk-rfcs/issues/431

closes #2809

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

----

Co-authored-by: Matt McClean <mmcclean@amazon.com>
Co-authored-by: Long Yao <yl1984108@gmail.com>
Co-authored-by: Drew Jetter <60628154+jetterdj@users.noreply.github.com>
Co-authored-by: Murali Ganesh <59461079+foxpro24@users.noreply.github.com>
Co-authored-by: Abilash Rangoju <988529+rangoju@users.noreply.github.com>

### CODE REVIEW

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+      role: this.makeScalingRole(),
+      minCapacity: scalingProps.minCapacity || this.initialInstanceCount,
+      maxCapacity: scalingProps.maxCapacity || this.initialInstanceCount,
+    });
+  }
+
+  /**
+   * Return the service linked role that will be used for AutoScaling
+   *
+   * Documentation is available here: https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-service-linked-roles.html
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:261]: The documentation here is good, and helped me find what I was looking for. But let's pull this into the docstring anyway. Application autoscaling creates service-linked roles for you, that is why they are being hardcoded here (maybe that's obvious for app autoscaling folks, but as a generic CDK guy i was unaware). 

"Return the pre-created service linked role..." or something

File: packages/@aws-cdk/aws-sagemaker/README.md [hunk 2]
```diff
@@ -195,3 +195,60 @@ const endpointConfig = new sagemaker.EndpointConfig(this, 'EndpointConfig', {
...
+more information about the API, see the
+[InvokeEndpoint](https://docs.aws.amazon.com/sagemaker/latest/dg/API_runtime_InvokeEndpoint.html)
+API. Defining an endpoint requires at minimum the associated endpoint configuration:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+declare const endpointConfig: sagemaker.EndpointConfig;
+
+const endpoint = new sagemaker.Endpoint(this, 'Endpoint', { endpointConfig });
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/README.md:213]: Instead of supplying an endpoint config construct, could you simply supply a raw config into `Endpoint`, and get `Endpoint` to spin up the `EndpointConfig` for you? What we're offering makes sense if you understand the use case of repeating endpointConfigs elsewhere, but may seem strange for someone who just wants one configuration on the endpoint. In that use case, it makes more sense to supply the raw config to `Endpoint` and then abstract away the `EndpointConfig` construct from the user entirely.

Maybe the resolution here is to add this as an issue and track +1s. If people want it, we can add it in later.

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    });
+  }
+}
+
+/**
+ * Construction properties for a SageMaker Endpoint.
+ */
+export interface EndpointProps {
+
+  /**
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:307]:
```suggestion
export interface EndpointProps {
  /**
```

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+ * The interface for a SageMaker Endpoint resource.
+ */
+export interface IEndpoint extends cdk.IResource {
+  /**
+   * The ARN of the endpoint.
+   *
+   * @attribute
+   */
+  readonly endpointArn: string;
+  /**
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:33]:
```suggestion
  readonly endpointArn: string;

  /**
```

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+ * Represents an instance production variant that has been associated with an endpoint.
+ */
+export interface IEndpointInstanceProductionVariant extends IEndpointProductionVariant {
+  /**
+   * Metric for the number of invocations
+   *
+   * @default - sum over 5 minutes
+   */
+  metricInvocations(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+  /**
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:74]: bigtime nit, can we add a space between all the methods listed here?

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+    });
+
+    class Import extends EndpointBase {
+      public endpointArn = endpointArn;
+      public endpointName = endpointName;
+    }
+
+    return new Import(scope, id, {
+      environmentFromArn: endpointArn,
+    });
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:379]: I don't like this duplication. We should have a third method, `fromEndpointAttributes` where the `Import` class lives, and both `fromEndpointArn` and `fromEndpointName` call that method.

File: packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    });
+  }
+
+  /**
+   * Calculate target value based on a ScalableProductionVariant
+   *
+   * Documentation for the equation is here: https://docs.aws.amazon.com/sagemaker/latest/dg/endpoint-scaling-loadtest.html
+   * @param scalableProductionVariant ScalableProductionVariant instance
+   */
+  private calculateScalingTarget( props: InvocationsScalingProps): number {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts:43]:
```suggestion
  private calculateScalingTarget(props: InvocationsScalingProps): number {
```

File: packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    const safetyFactor = props.safetyFactor || 0.5;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts:44]: are there any sane guardrails we should be putting on the safetyFactor? Is there a use case for someone to make `safetyFactor > 1`? Or 0? A few of these edge cases would result in `calculateScalingTarget` returning some problematic info, whether thats ridiculously high or low.

File: packages/@aws-cdk/aws-sagemaker/README.md [hunk 9]
```diff
@@ -195,3 +195,60 @@ const endpointConfig = new sagemaker.EndpointConfig(this, 'EndpointConfig', {
...
+
+To enable autoscaling on the production variant, use the `autoScaleInstanceCount` method:
+
+```typescript
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+declare const endpointConfig: sagemaker.EndpointConfig;
+
+const endpoint = new sagemaker.Endpoint(this, 'Endpoint', { endpointConfig });
+const productionVariant = endpoint.findInstanceProductionVariant('variantName');
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/README.md:226]: how would you know the variant name here? Is it defined as part of `endpointConfig`? If so, I think it's worth defining the endpointConfig completely and pulling `variantName` out into a variable, at least for this example so it's a bit more straightforward.

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+import { ScalableInstanceCount } from './scalable-instance-count';
+
+/*
+ * Amazon SageMaker automatic scaling doesn't support automatic scaling for burstable instances such
+ * as T2, because they already allow for increased capacity under increased workloads.
+ * https://docs.aws.amazon.com/sagemaker/latest/dg/endpoint-auto-scaling-add-console.html
+ */
+const BURSTABLE_INSTANCE_TYPE_PREFIXES = Object.entries(ec2.InstanceClass)
+  .filter(([name, _]) => name.startsWith('T'))
+  .map(([_, prefix]) => `ml.${prefix}.`);
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:21]: So I see `T3`, `T3a`, `T2`, and `T4G` are the burstable instances that do not support automatic scaling from your link. And here we are assuming that all `T` prefixed instances are burstable. I know nothing about ec2 instances so I'm going to take your word for it, but it would be a problem if AWS released `T5` tomorrow that wasn't burstable... I'm assuming thats unlikely

File: packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * The properties of a scalable attribute representing task count.
+ */
+export interface ScalableInstanceCountProps extends appscaling.BaseScalableAttributeProps {
+}
+
+/**
+ * A scalable sagemaker endpoint attribute
+ */
+export class ScalableInstanceCount extends appscaling.BaseScalableAttribute {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts:13]: This should probably be its own PR that highlights the feature, like `feat(sagemaker): autoscaling endpoints`. If it's not hard to do, I support splitting it up. But given that I've already reviewed, I'm ok with pushing forward as is too.

**Review by kaizencc — DISMISSED** [review]
Okidoke @petermeansrock! I reviewed most of the code and left the tests alone for now. Few minor comments like always.

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 2 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/README.md:213]: > Instead of supplying an endpoint config construct, could you simply supply **a raw config** into `Endpoint`, and get `Endpoint` to spin up the `EndpointConfig` for you?

1. What do you mean by "_a raw config_"?
    1. Do you mean take each of [these members of `EndpointConfigProps`](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-sagemaker/lib/endpoint-config.ts#L132-L145) and add them to `EndpointProps` as optional arguments (while making the `IEndpointConfig` prop value optional as well)?
    1. Or possibly, take `EndpointConfigProps` as a parameter on `EndpointProps`?
1. In this scenario, would it also make sense to include an `addInstanceProductionVariant` on `Endpoint` so that customers can mutate the backing `EndpointConfig` in the same manner as they could using their own instance (this does have the added bonus of returning an `IEndpointInstanceProductionVariant` instance from the method, eliminating the need for customers to look up the variant by name).
    1. If so, should this mutation only be permitted when creating the `EndpointConfig` on behalf of customers, or should it also be supported in cases when they provide their own `EndpointConfig` instance?

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts (hunk 6 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:379]: Should I mark this `fromEndpointAttributes` method as private? I ask because [the following design guideline](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#from-attributes) implies that the method should only be exposed for non-name, non-ARN scenarios:

> If a resource has more than a single attribute (“ARN” and “name” are usually considered a single attribute since it's usually possible to convert one to the other), then the resource should provide a static fromAttributes method to allow users to explicitly supply values to all resource attributes when they import an external (unowned) resource [awslint:from-attributes].

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 9 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/README.md:226]: > how would you know the variant name here? Is it defined as part of `endpointConfig`? If so, I think it's worth defining the endpointConfig completely and pulling `variantName` out into a variable, at least for this example so it's a bit more straightforward.

Correct, and that feedback sounds good. Should I follow the same pattern in the "_Metrics_" section since it too performs `findInstanceProductionVariant`?

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts (hunk 10 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:21]: The following snippet of [EC2 documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/burstable-performance-instances.html) implies that all `T` instances will be burstable:

> These low-to-moderate CPU utilization workloads lead to wastage of CPU cycles and, as a result, you pay for more than you use. To overcome this, you can leverage the low-cost burstable general purpose instances, which are the T instances.
> 
> The T instance family provides a baseline CPU performance with the ability to burst above the baseline at any time for as long as required.

File: packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts (hunk 11 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/lib/scalable-instance-count.ts:13]: > But given that I've already reviewed, I'm ok with pushing forward as is too.

Since you've already reviewed these pieces, I'd prefer to keep them as part of this PR (it saves me a few cycles on excising the feature and regenerating integ test snapshots). As part of subsequent revisions on this PR though, if you begin to feel more strongly that auto-scaling should be split apart, I can certainly do it.

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 9 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-sagemaker/README.md:226]: I'm ok with it just being here. This example focuses on a method exposed on the `productionVariant`, so it makes sense to be a little more explicit about how we're getting the variant.

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 2 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-sagemaker/README.md:213]: I am tossing 1i. out as the suggestion, along with ensuring that the "raw config" and the `IEndpointConfig` prop are mutually exclusive. 

However, this is one of those things that is easier to add and harder to take away. So I think this should be a no-op for now, and we wait to see if something like this is a common complaint amongst users.

File: packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts (hunk 6 above)
Comment (kaizencc) [reply packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts:379]: I'm reading that design guideline as `fromXxxAttributes` should be added if a resource has more than a single attribute, but not specifically disallowing the method when thats not the case. The reason for this linter rule is likely to guard against the duplication we have here when we have multiple `fromXxxThing` methods.

I am for adding in a public `fromXxAttributes`, unless our linter rules somehow flag that as wrong, which would surprise me.

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 2 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/README.md:213]: Sounds good. Once this PR gets merged, would you like me to proactively create an issue to track that feature request?

Comment (kaizencc) [reply packages/@aws-cdk/aws-sagemaker/README.md:213]: Yep, that would be great :)

File: packages/@aws-cdk/aws-sagemaker/test/integ.endpoint.ts [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+  maxCapacity: 3,
+});
+instanceCount.scaleOnInvocations('LimitRPS', {
+  maxRequestsPerSecond: 30,
+});
+
+productionVariant.metricInvocations().createAlarm(stack, 'InvocationsAlarm', {
+  threshold: 1,
+  evaluationPeriods: 2,
+});
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-sagemaker/test/integ.endpoint.ts:87]: last thing... these alarms don't seem to be needed since we're not asserting that they exist or anything. Lets move this to a separate integ test as it doesn't really test the endpoint construct. maybe `integ.alarms.ts` or something.

It is cluttering the current integ test atm.

**Review by kaizencc — DISMISSED** [review]
Adding changes requested since I'm looking for one thing in the integ test from an earlier comment. Everything else looks good, feel bad for blocking the approval on one thing but hopefully that's not too troublesome... once the integ tests are split out, I'm ready to slap the approval. Thanks for these quality PRs @petermeansrock!

**Review by kaizencc — APPROVED** [review]
Yay @petermeansrock, thanks for all your hard work and quick turn around time! This looks good to me, and means that your RFC design is fully implemented. Congrats! If you are willing/interested, it will be interesting to see what kind of usage + issues we get on this new L2 library and would be great to have your continued participation on those issues as they come in! I may try to ping you...

File: packages/@aws-cdk/aws-sagemaker/README.md (hunk 2 above)
Comment (petermeansrock) [reply packages/@aws-cdk/aws-sagemaker/README.md:213]: Closing the loop on this thread, I've created the following issue to track interest in this feature:
* #23162
